dissect/cstruct/types/structure.py

```diff
@@ -65,29 +77,39 @@ def _update_fields(
 
         lookup = {}
         raw_lookup = {}
+        init_names = []
+        field_names = []
         for field in fields:
             if field.name in lookup and field.name != "_":
                 raise ValueError(f"Duplicate field name: {field.name}")
 
-            if isinstance(field.type, StructureMetaType) and field.type.anonymous:
-                lookup.update(field.type.lookup)
+            if isinstance(field.type, StructureMetaType) and field.type.__anonymous__:
+                for anon_field in field.type.fields.values():
+                    attr = f"{field.name}.{anon_field.name}"
+                    classdict[anon_field.name] = property(attrgetter(attr), attrsetter(attr))
+
+                lookup.update(field.type.fields)
             else:
                 lookup[field.name] = field
 
             raw_lookup[field.name] = field
 
         num_fields = len(lookup)
         field_names = lookup.keys()
-        classdict["fields"] = fields
-        classdict["lookup"] = lookup
-        classdict["__lookup__"] = raw_lookup
-        classdict["__init__"] = _patch_args_and_attributes(_make__init__(num_fields), field_names)
+        init_names = raw_lookup.keys()
+        classdict["fields"] = lookup
+        classdict["lookup"] = raw_lookup
+        classdict["__fields__"] = fields
         classdict["__bool__"] = _patch_attributes(_make__bool__(num_fields), field_names, 1)
 
-        if issubclass(cls, UnionMetaType):
+        if issubclass(cls, UnionMetaType) or isinstance(cls, UnionMetaType):
```

Comment:
```suggestion
        if issubclass(cls, UnionMetaType) or isinstance(cls, UnionMetaType):
```

`issubclass` will raise an exception if `cls` is an instance, so this needs to be guarded.

Reply:
`cls` is always a class. The `isinstance()` here checks if it's an instance of a `UnionMetaType`, which is a class.

Reply:
Ah, because it is part of the metaclass apparently, check.

Reply:
I had to double check as well, but this code path has unit tests for both conditions so I can at least assure it works :smile: